python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -323,40 +332,82 @@ def _messages_thread(
         self,
         handler: "PipesMessageHandler",
         params: PipesParams,
-        is_task_complete: Event,
+        is_session_closed: Event,
     ) -> None:
         start_or_last_download = datetime.datetime.now()
         while True:
             now = datetime.datetime.now()
-            if (now - start_or_last_download).seconds > self.interval or is_task_complete.is_set():
+            if (now - start_or_last_download).seconds > self.interval or is_session_closed.is_set():
                 start_or_last_download = now
                 chunk = self.download_messages_chunk(self.counter, params)
                 if chunk:
                     for line in chunk.split("\n"):
                         message = json.loads(line)
                         handler.handle_message(message)
                     self.counter += 1
-                elif is_task_complete.is_set():
+                elif is_session_closed.is_set():
+                    break
+            time.sleep(DEFAULT_SLEEP_INTERVAL)
+
+    def _logs_thread(
+        self,
+        params: PipesParams,
+        is_session_closed: Event,
+    ) -> None:
+        # Start once we have received the opened message from the external process
+        while self.opened_payload is None:
+            time.sleep(DEFAULT_SLEEP_INTERVAL)
```

Comment:
we should also exit this while loop on session close event

Reply:
It's tricky-- can't quite do that because the `is_session_closed` could be set before the `opened_payload` exists, since reading the `opened_payload` happens in a different thread. And we can't use another timeout here because we don't know how long it will take for the external process to even start executing.

If the `opened_payload` is never received then what will happen right now is that the thread timeout will trigger an error from this never exiting. Open to other ideas.

Reply:
I'm hesitant to introduce another event but maybe we want e.g. `are_all_messages_processed`, that we could use.

Reply:
> can't quite do that because the is_session_closed could be set before the opened_payload exists

Ya as the code youve added against the new event does, we just need to bail the thread if session closes before open payload

Reply:
>we just need to bail the thread if session closes before open payload

Unclear to me whether you think more needs to be done here now-- after latest iter we will bail if all messages have been processed and no opened is received, but _not_ if `is_session_closed` is set and no opened is received (yet), which is correct I think.